app/src/main/java/org/mozilla/fenix/components/metrics/AppLaunchTimeMeasurement.kt

```diff
@@ -0,0 +1,78 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.os.Process
+import android.os.SystemClock
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.mozilla.fenix.perf.Stat
+
+/**
+ * Handles the logic of figuring out launch time for cold, warm and hot startup.
+ */
+class AppLaunchTimeMeasurement(private val stats: Stat = Stat()) {
+    private var isColdStartup = false
+    private var isWarmStartup = false
+    private var isHotStartup = false
+    private var isOnPreDrawCalled = false
+
+    private var applicationOnCreateTimeStampNanoSeconds: Long? = null
+    private var homeActivityInitTimeStampNanoSeconds: Long? = null
+    private var homeActivityOnRestartTimeStampNanoSeconds: Long? = null
+    // we are considering screen to be visible after the first pre draw call.
+    private var homeActivityOnPreDrawTimeStampNanoSeconds: Long = INVALID_TIMESTAMP
+
+    fun onApplicationOnCreate() {
+        isColdStartup = true
+    }
+
+    fun onHomeActivityOnCreate(currentTimeStampNanoSeconds: Long) {
+        if (!isColdStartup) {
+            isWarmStartup = true
+        }
+        this.homeActivityInitTimeStampNanoSeconds = currentTimeStampNanoSeconds
+    }
+
+    fun onHomeActivityOnRestart(currentTimeStampNanoSeconds: Long = SystemClock.elapsedRealtimeNanos()) {
```

Comment:
We should capture this time before `super.onRestart`, right? It might take a non-trivial amount of time to run that and other folks might add code in between that and this.

Either call this first or capture the time first and pass it to this method.